src/services/organizeImports.ts

```diff
@@ -0,0 +1,291 @@
+/* @internal */
+namespace ts.OrganizeImports {
+    export function organizeImports(
+        sourceFile: SourceFile,
+        formatContext: formatting.FormatContext,
+        host: LanguageServiceHost,
+        cancellationToken: CancellationToken) {
+
+        // All of the (old) ImportDeclarations in the file, in syntactic order.
+        const oldImportDecls: ImportDeclaration[] = [];
+
+        forEachChild(sourceFile, node => {
+            cancellationToken.throwIfCancellationRequested();
+            if (isImportDeclaration(node)) {
+                oldImportDecls.push(node);
+            }
+            // TODO (https://github.com/Microsoft/TypeScript/issues/10020): sort *within* ambient modules (find using isAmbientModule)
+        });
+
+        if (oldImportDecls.length === 0) {
+            return [];
+        }
+
+        const usedImportDecls = removeUnusedImports(oldImportDecls);

```

Comment:
Probably too many calls to `throwIfCancellationRequested` here - in general you only need to call this every ~50-200ms, and I don't think `sortImports` or `coalesceImports` could ever be that slow.